packages/starksheet-webapp/src/provider/EVMProvider.ts

```diff
@@ -190,31 +192,75 @@ export class EVMProvider implements ChainProvider {
   execute = async (calls: Call[], options: { value: number | string })  => {
     const provider = new ethers.providers.Web3Provider(window.ethereum);
     const signer = provider.getSigner();
+    const MULTISEND_CONTRACT_ADDRESS = process.env.REACT_APP_MULTISEND_ADDRESS || "";
 
+    const encodeTransactions = await Promise.all(calls.map(async (call, index) => {
+      let abi = await this.getAbi(call.contractAddress);
+      if(abi.length === 0) {
+        abi = evmWorksheetAbi;
+      }
+      // todo: implement cleanly
 
-    // Callers of the execute function assume the chain supports multi-call, which EVM does not natively.
-    // The following is a hack: we wait for every transaction before resolving, then returning one of the transactions.
-    // This works because callers do not really care about the transaction response, only its status and completion
-    const receipts = await Promise.all(calls.map(async (call) => {
-      const abi = await this.getAbi(call.contractAddress);
       const contract = new Contract(call.contractAddress, abi, signer);
 
-      const response: ethers.providers.TransactionResponse = await contract[call.entrypoint](...(call.calldata || []), {
-        value: options.value,
-      });
-      return await response.wait();
-      }));
+      const fragment = contract.interface.fragments.find((fragment) => fragment.name === call.entrypoint);
 
-    const firstToFail = receipts.find((receipt) => receipt.status !== 1)
-    if (firstToFail) {
-      return {
-        transaction_hash: firstToFail.transactionHash,
+      if (!fragment) {
+        throw new Error(`Could not find fragment for ${call.entrypoint} entrypoint in contract ABI`);
+      }
+      const signature = fragment.format();
+  
+      const inputValues= {
+        0: ""
       };
+      (call.calldata as RawCalldata).forEach((value, index) => {
+        inputValues[index as keyof typeof inputValues] = value as string;
+      });
+      
+      const transactionInput : CallContractTransactionInput = {
+        type: TransactionType.callContract,
+        id: index.toString(),
+        to: call.contractAddress,
+        value: call.entrypoint === "addSheet" ? options.value.toString() : "0",
+        abi: abi,
+        functionSignature: signature,
+        inputValues
+      }
+  
+      const metaTransaction = encodeSingle(transactionInput);
+      
+      return metaTransaction;
+    }));
+    
+    const transactions = encodeMulti(encodeTransactions, MULTISEND_CONTRACT_ADDRESS);
+    const multiSendTx = ethers.utils.solidityPack(
+      ["uint8", "address", "uint256", "uint256", "bytes"],
+      [0, transactions.to, 0, transactions.data.length, transactions.data]
+    );
+
+    let isPayable = false;
+    for(let i = 0; i < calls.length; i++) {
+      if(calls[i].entrypoint === "addSheet") {
+        isPayable = true;
+        break;
+      } 
     }
```

Comment:
```suggestion
```

do we need this ? we should rely on the method caller to set the value, so the check feels not usefull, as well as not generic